File: apple/Sources/CCDAEngine/Models/Entry/CCDAEntryStatus.swift
import Foundation

/// Status code attached to a structured C-CDA entry.
public enum CCDAEntryStatus: Hashable, Sendable {
    /// Entry is currently active.
    case active
    /// Entry has been completed.
    case completed
    /// Entry was aborted before completion.
    case aborted
    /// Entry was cancelled.
    case cancelled
    /// Entry is temporarily suspended.
    case suspended
    /// Entry is temporarily held.
    case held
    /// Entry is newly created or pending.
    case new
    /// Entry status is normal.
    case normal
    /// Entry has been made obsolete.
    case obsolete
    /// Entry has been nullified.
    case nullified
    /// Unsupported status code preserved from source XML.
    case unsupported(String)

    /// Creates an entry status from a C-CDA `statusCode/@code` value.
    public init(code: String) {
        switch code {
        case "active":
            self = .active
        case "completed":
            self = .completed
        case "aborted":
            self = .aborted
        case "cancelled":
            self = .cancelled
        case "suspended":
            self = .suspended
        case "held":
            self = .held
        case "new":
            self = .new
        case "normal":
            self = .normal
        case "obsolete":
            self = .obsolete
        case "nullified":
            self = .nullified
        default:
            self = .unsupported(code)
        }
    }

    /// Original C-CDA status code.
    public var rawValue: String {
        switch self {
        case .active:
            return "active"
        case .completed:
            return "completed"
        case .aborted:
            return "aborted"
        case .cancelled:
            return "cancelled"
        case .suspended:
            return "suspended"
        case .held:
            return "held"
        case .new:
            return "new"
        case .normal:
            return "normal"
        case .obsolete:
            return "obsolete"
        case .nullified:
            return "nullified"
        case .unsupported(let rawValue):
            return rawValue
        }
    }
}

File: android/ccda-engine/src/main/kotlin/com/ccdakit/engine/Models/Entry/CCDAEntryStatus.kt
package com.ccdakit.engine

/** Status code attached to a structured C-CDA entry. */
sealed class CCDAEntryStatus(open val rawValue: String) {
    /** Entry is currently active. */
    data object Active : CCDAEntryStatus("active")
    /** Entry has been completed. */
    data object Completed : CCDAEntryStatus("completed")
    /** Entry was aborted before completion. */
    data object Aborted : CCDAEntryStatus("aborted")
    /** Entry was cancelled. */
    data object Cancelled : CCDAEntryStatus("cancelled")
    /** Entry is temporarily suspended. */
    data object Suspended : CCDAEntryStatus("suspended")
    /** Entry is temporarily held. */
    data object Held : CCDAEntryStatus("held")
    /** Entry is newly created or pending. */
    data object New : CCDAEntryStatus("new")
    /** Entry status is normal. */
    data object Normal : CCDAEntryStatus("normal")
    /** Entry has been made obsolete. */
    data object Obsolete : CCDAEntryStatus("obsolete")
    /** Entry has been nullified. */
    data object Nullified : CCDAEntryStatus("nullified")
    /** Unsupported status code preserved from source XML. */
    data class Unsupported(override val rawValue: String) : CCDAEntryStatus(rawValue)

    companion object {
        /** Creates an entry status from a C-CDA `statusCode/@code` value. */
        fun fromCode(code: String): CCDAEntryStatus = when (code) {
            "active" -> Active
            "completed" -> Completed
            "aborted" -> Aborted
            "cancelled" -> Cancelled
            "suspended" -> Suspended
            "held" -> Held
            "new" -> New
            "normal" -> Normal
            "obsolete" -> Obsolete
            "nullified" -> Nullified
            else -> Unsupported(code)
        }
    }
}
